tools/QuietBackgroundProcesses/DevHome.QuietBackgroundProcesses.PerformanceRecorderEngine/PerformanceRecorderEngine.cpp

```diff
@@ -0,0 +1,669 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+
+#include <chrono>
+#include <filesystem>
+#include <iostream>
+#include <map>
+#include <mutex>
+#include <set>
+#include <span>
+#include <string>
+#include <string_view>
+#include <thread>
+#include <vector>
+
+#include <wil/resource.h>
+#include <wil/token_helpers.h>
+#include <wil/win32_helpers.h>
+
+#include <windows.h>
+#include <appmodel.h>
+#include <psapi.h>
+
+#include "PerformanceRecorderEngine.h"
+
+// 2 percent is the threshold for a process to be considered as a high CPU consumer
+#define CPU_TIME_ABOVE_THRESHOLD_STRIKE_VALUE 0.02f
+
+enum class ProcessCategory
+{
+    Unknown,
+    User,
+    System,
+    Developer,
+    Background,
+};
+
+struct ProcessPerformanceInfo
+{
+    // Process id
+    wil::unique_process_handle process;
+    ULONG pid{};
+
+    // Process info
+    std::wstring name;
+    std::wstring path;
+    std::optional<std::wstring> packageFullName;
+    std::optional<std::wstring> aumid;
+    ProcessCategory category{};
+    FILETIME createTime{};
+    FILETIME exitTime{};
+
+    // CPU times
+    FILETIME startUserTime{};
+    FILETIME startKernelTime{};
+    FILETIME previousUserTime{};
+    FILETIME currentUserTime{};
+    FILETIME previousKernelTime{};
+    FILETIME currentKernelTime{};
+
+    // Sampling
+    uint64_t sampleCount{};
+    float percentCumulative{};
+    float varianceCumulative{};
+    float sigma4Cumulative{};
+    float maxPercent{};
+    uint32_t samplesAboveThreshold{};
+};
+
+// Process Categories
+std::set<std::wstring> c_user = {
+    L"chrome.exe",
```

Comment:
Sure it is a good idea to hard code these values?

Reply:
This is one of those things I know I should do, but I really don't want to.

Do you have any advice?

Reply:
I do not.

Reply:
JSON seems like a lot of work to read.  Could just do .txt files, but then I need 5 of them.